pypst/utils.py

```diff
@@ -62,3 +101,153 @@ def render_sequence(arg: Iterable[Any]) -> str:
     Render a sequence of any object supported by `render`.
     """
     return f"({', '.join(render_code(a) for a in arg)})"
+
+
+def render_datetime(arg: date | datetime) -> str:
+    """
+    Render a Python `datetime.date` into a call to the Typst datetime function.
+    """
+    obj = {
+        name: getattr(arg, name)
+        for name in ["year", "month", "day", "hour", "minute", "second"]
+        if hasattr(arg, name)
+    }
+    return f"#datetime{render_mapping(obj)}"
+
+
+def render_timedelta(arg: timedelta) -> str:
+    """
+    Render a Python `datetime.timedelta` into a call to the Typst duration function.
+    """
+    obj = {
+        name: getattr(arg, name)
+        for name in [
+            "microseconds",
+            "seconds",
+            "days",
+        ]
+        if hasattr(arg, name)
+    }
+    obj["seconds"] = obj["seconds"] + round(obj.pop("microseconds") / 1e6)
+    return f"#duration{render_mapping(obj)}"
+
+
+@dataclass
+class Dictionary:
+    """
+    Helper class to render Python dataclasses by iterating over its fields
+    as if it were a dictionary.
+
+    Inherit from `Dictionary` to inherit the render method
+    and use it with your dataclass.
+
+    You can specify that `None` values should be rendered instead
+    of being skipped with `keep_none=True` on the field's metadata.
+
+    Example:
+        >>> from dataclasses import dataclass, field
+        >>> @dataclass
+        ... class Foo(Dictionary):
+        ...    bar: int | None = 4
+        ...    qux: int | None = field(default=16, metadata={"keep_none": True})
+        >>>
+        >>> Foo().render()
+        '#(bar: 4, qux: 16)'
+        >>> Foo(bar=None).render()
+        '#(qux: 16)'
+        >>> Foo(qux=None).render()
+        '#(bar: 4, qux: none)'
+    """
+
+    def __str__(self) -> str:
+        return self.render()
+
+    def render(self) -> str:
+        """
+        Dataclass rendering to a Typst dictionary.
+
+        Modify `fields_to_render` to control which fields are included.
+        """
+        mapping = render_mapping(
+            {field.name: getattr(self, field.name) for field in self.fields_to_render()}
+        )
+        return f"#{mapping}"
+
+    def fields_to_render(self) -> Iterable[Field]:
+        """
+        These fields should be rendered. Defaults to skipping any fields with a value of `None`.
+        """
+
+        def check(field: Field) -> bool:
+            # When the value is `None`, check whether that should be kept.
+            if getattr(self, field.name) is None:
+                return field.metadata.get("keep_none", False)
+
+            # Keep otherwise.
+            return True
+
+        return filter(check, fields(self))
```

Comment:
Nice addition, but I think I'm not on board with the specific implementation. As dataclasses are only treated as mappings, I think we should integrate rendering them in the `render` function. This way, we avoid the inheritance which may get complicated if the user has its own inheritance structure. The `fields_to_render` function can exist as a stand-alone utility function that expects a dataclass as its argument.